pkg/tfbridge/detailed_diff.go

```diff
@@ -41,6 +41,105 @@ func sortedMergedKeys[K cmp.Ordered, V any, M ~map[K]V](a, b M) []K {
 	return keysSlice
 }
 
+func isTypeShapeMismatched(val resource.PropertyValue, propType shim.ValueType) bool {
+	if !isPresent(val) {
+		return false
+	}
+	switch propType {
+	case shim.TypeList:
+		return !val.IsArray()
+	case shim.TypeSet:
+		return !val.IsArray()
+	case shim.TypeMap:
+		return !val.IsObject()
+	default:
+		return false
+	}
+}
+
+func lookupSchemas(
+	path propertyPath, tfs shim.SchemaMap, ps map[string]*info.Schema,
+) (shim.Schema, *info.Schema, error) {
+	schemaPath := PropertyPathToSchemaPath(resource.PropertyPath(path), tfs, ps)
+	return LookupSchemas(schemaPath, tfs, ps)
+}
+
+// walkPropertyValue walks a property value and calls the visitor function for each path in the property value.
+func walkPropertyValue(
+	val resource.PropertyValue, path propertyPath, visitor func(propertyPath, resource.PropertyValue) bool,
+) bool {
+	if !visitor(path, val) {
+		return false
+	}
+
+	switch {
+	case val.IsArray():
+		for i, elVal := range val.ArrayValue() {
+			if !walkPropertyValue(elVal, path.Index(i), visitor) {
+				return false
+			}
+		}
+
+	case val.IsObject():
+		for k, elVal := range val.ObjectValue() {
+			if !walkPropertyValue(elVal, path.Subpath(string(k)), visitor) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func willTriggerReplacement(
+	path propertyPath, rootTFSchema shim.SchemaMap, rootPulumiSchema map[string]*info.Schema,
+) bool {
+	// A change on a property might trigger a replacement if:
```

Comment:
The function seems to do what the comment says. The rule sounds a little un-intuitive but looks like you've explored this thoroughly in tests so perhaps that's that? 

If a property P is forcenew, but P.a.b.c.d has changed, I am guessing that the overall plan is a replace plan, is that right? If we are displaying the detailed diff as `a.b.c.d has changed` we are probably within rights to blame `d` for replacing?

Reply:
That's not quite the case as any object along the way stops the propagation of the change.

in your example if b is an object and d has changed then TF will list b as updated instead and not trigger a replace on P.

See also https://github.com/pulumi/pulumi-terraform-bridge/pull/2496#discussion_r1819350356

Cross tests for this in https://github.com/pulumi/pulumi-terraform-bridge/blob/49aaf757dc71785b2dd60ee0af249f794f19235f/pkg/internal/tests/cross-tests/diff_cross_test.go#L1369